hail/hail/src/is/hail/backend/service/Worker.scala

```diff
@@ -76,141 +75,198 @@ class ExplicitClassLoaderInputStream(is: InputStream, cl: ClassLoader)
+      val bs = new Array[Byte](length)
+      is.readFully(bs)
+      bs
+    }
+
+    def readString(): String =
+      new String(is.readBytes(), StandardCharsets.UTF_8)
+
+    def readSuccess(): (Array[Byte], Int) = {
+      val partId = is.readInt()
+      val bytes = is.readBytes()
+      bytes -> partId
+    }
+
+    def readFailure(): HailWorkerException =
+      HailWorkerException(
+        partitionId = is.readInt(),
+        shortMessage = is.readString(),
+        expandedMessage = is.readString(),
+        errorId = is.readInt(),
+      )
   }
 
-  def main(argv: Array[String]): Unit = {
-    val theHailClassLoader = new HailClassLoader(getClass().getClassLoader())
+  def write(os: DataOutputStream, partId: Int, result: Either[Throwable, Array[Byte]]): Unit =
+    result match {
+      case Left(throwable) =>
+        os.writeByte(0)
+        os.writeFailure(partId, throwable)
+
+      case Right(bytes) =>
+        os.writeByte(1)
+        os.writeSuccess(partId, bytes)
+    }
+
+  def read(is: DataInputStream): Either[HailWorkerException, (Array[Byte], Int)] =
+    is.readByte() match {
+      case 0 => Left(is.readFailure())
+      case 1 => Right(is.readSuccess())
+    }
+}
 
+object FutureOps {
+  // "unsafe" in that sequencing Future[Lazy[_]] does not guarantee completion order
+  def unsafeInterleave[A](f: => A)(implicit ec: ExecutionContext): Future[Lazy[A]] =
+    Future {
+      val fut = Future(blocking(f))
+      lazily[A](Await.result(fut, Duration.Inf))
+    }
+}
+
+object Worker extends Logging {
+
+  def main(argv: Array[String]): Unit = {
     if (argv.length != 7) {
       throw new IllegalArgumentException(s"expected seven arguments, not: ${argv.length}")
     }
+
     val scratchDir = argv(0)
     // val logFile = argv(1)
     // var jarLocation = argv(2)
     val kind = argv(3)
     assert(kind == Main.WORKER)
     val root = argv(4)
-    val i = argv(5).toInt
-    val n = argv(6).toInt
+    val partition = argv(5).toInt
+    val index = argv(6).toInt
     val timer = new WorkerTimer()
 
     val deployConfig = DeployConfig.fromConfigFile("/deploy-config/deploy-config.json")
     DeployConfig.set(deployConfig)
     sys.env.get("HAIL_SSL_CONFIG_DIR").foreach(tls.setSSLConfigFromDir)
 
-    log.info(s"is.hail.backend.service.Worker $HAIL_REVISION")
-    log.info(s"running job $i/$n at root $root with scratch directory '$scratchDir'")
+    log.info(s"${getClass.getName} $HAIL_REVISION")
+    log.info(s"running partition $partition root '$root' with scratch directory '$scratchDir'")
+
+    timer.start(s"partition $partition")
 
-    timer.start(s"Job $i/$n")
+    val hcl = new HailClassLoader(getClass.getClassLoader)
 
-    timer.start("readInputs")
     val fs = RouterFS.buildRoutes(
       CloudStorageFSConfig.fromFlagsAndEnv(
         Some(Path.of(scratchDir, "secrets/gsa-key/key.json")),
         HailFeatureFlags.fromEnv(),
       )
     )
 
+    import ExecutionContext.Implicits.global
```

Comment:
A few observations here:
- we were never properly managing the lifetime of the underlying `ExecutorService` 
- we were not doing anything fancy with this or doing any real customisation of number of thread etc
- I don't understand our `ClassLoader` caching and juggling: scala objects are only unloaded when the class loader that loads them is unloaded. Yet `JVMEntryway` caches classloaders. Are these objects then ever unloaded?


Given that, why not just use scala's global execute context?

Reply:
The scala docs [here](https://docs.scala-lang.org/overviews/core/futures.html#execution-context) suggest cases where you'd want to use a separate executor, but I don't understand well enough to know if that applies here.